File: viewStocks/backend/src/main/java/hu/elte/alkfejl/Stocks/viewStocks/service/PortfolioService.java
package hu.elte.alkfejl.Stocks.viewStocks.service;

import hu.elte.alkfejl.Stocks.viewStocks.model.Portfolio;
import hu.elte.alkfejl.Stocks.viewStocks.model.Transaction;
import hu.elte.alkfejl.Stocks.viewStocks.repository.PortfolioRepository;
import hu.elte.alkfejl.Stocks.viewStocks.wrapper.PortfolioDto;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.SessionScope;

import java.util.ArrayList;
import java.util.List;

@Service
@SessionScope
public class PortfolioService {

    @Autowired
    private PortfolioRepository portfolioRepository;

    public Portfolio add(Portfolio portfolio) {
        return portfolioRepository.save(portfolio);
    }

    public void delete(Portfolio portfolio) {
        portfolioRepository.delete(portfolio);
    }

    public Portfolio update(Portfolio portfolio, Transaction transaction, String ticker) {
        if(portfolio.getTransactions().remove(transaction)) {
            transaction.setTicker(ticker);
            portfolio.getTransactions().add(transaction);
        }
        return portfolioRepository.save(portfolio);
    }

    public Portfolio findById(Long portfolioId) {
        return portfolioRepository.findOne(portfolioId);
    }

    public List<PortfolioDto> getOwnedPortfolios(Long userId) {
        List<Portfolio> portfolioList = portfolioRepository.findByOwnerId(userId);
        List<PortfolioDto> portfolioDtoList = new ArrayList<>();

        for (Portfolio p : portfolioList) {
            portfolioDtoList.add(new PortfolioDto(p));
        }

        return portfolioDtoList;
    }
}

File: viewStocks/frontend/src/app/_services/portfolio.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {URL} from '../_constants';
import {Portfolio} from '../_models';

@Injectable()
export class PortfolioService {
  constructor(private http: HttpClient) {
  }

  add(portfolio: Portfolio) {
    return this.http.post(`${URL}/portfolio/add`, portfolio);
  }

  delete(portfolio: Portfolio) {
    // return this.http.delete(`${URL}/portfolio/delete`, portfolio);
  }

  update(portfolio: Portfolio) {
    return this.http.put(`${URL}/portfolio/update`, portfolio);
  }

  findById(portfolio: Portfolio) {
    return this.http.get(`${URL}/portfolio/get?portfolioId=${portfolio.id}`);
  }

  getOwnedPortfolios(portfolio: Portfolio) {
    return this.http.get(`${URL}/portfolio/getOwnedPortfolios?userId=${portfolio.owner.id}`);
  }

}
